src/main/drivers/compass/compass_ak8963.c

```diff
@@ -127,83 +175,53 @@ static uint32_t ak8963SensorQueuedReadTimeRemaining(void)
 }
 
-static bool ak8963SensorCompleteRead(uint8_t *buf)
+static bool ak8963SlaveCompleteRead(const busDevice_t *slavedev, uint8_t *buf)
 {
-    uint32_t timeRemaining = ak8963SensorQueuedReadTimeRemaining();
+    uint32_t timeRemaining = ak8963SlaveQueuedReadTimeRemaining();
+
+    const busDevice_t *bus = slavedev->busdev_u.i2c.master;
 
     if (timeRemaining > 0) {
         delayMicroseconds(timeRemaining);
     }
 
     queuedRead.waiting = false;
 
-    spiBusReadRegisterBuffer(bus, MPU_RA_EXT_SENS_DATA_00, buf, queuedRead.len);               // read I2C buffer
+    spiBusReadRegisterBuffer(bus, MPU_RA_EXT_SENS_DATA_00, buf, queuedRead.len);    // read I2C buffer
     return true;
 }
-#else
-static bool ak8963SensorRead(uint8_t addr_, uint8_t reg_, uint8_t len, uint8_t* buf)
-{
-    return i2cRead(MAG_I2C_INSTANCE, addr_, reg_, len, buf);
-}
-
-static bool ak8963SensorWrite(uint8_t addr_, uint8_t reg_, uint8_t data)
-{
-    return i2cWrite(MAG_I2C_INSTANCE, addr_, reg_, data);
-}
-#endif
 
-static bool ak8963Init(void)
+static bool ak8963SlaveReadData(const busDevice_t *busdev, uint8_t *buf)
 {
-    uint8_t calibration[3];
-    uint8_t status;
-
-    ak8963SensorWrite(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_CNTL1, CNTL1_MODE_POWER_DOWN); // power down before entering fuse mode
-    ak8963SensorWrite(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_CNTL1, CNTL1_MODE_FUSE_ROM); // Enter Fuse ROM access mode
-    ak8963SensorRead(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_ASAX, sizeof(calibration), calibration); // Read the x-, y-, and z-axis calibration values
+    typedef enum {
+        CHECK_STATUS = 0,
+        WAITING_FOR_STATUS,
+        WAITING_FOR_DATA
+    } ak8963ReadState_e;
 
-    magGain[X] = ((((float)(int8_t)calibration[X] - 128) / 256) + 1) * 30;
-    magGain[Y] = ((((float)(int8_t)calibration[Y] - 128) / 256) + 1) * 30;
-    magGain[Z] = ((((float)(int8_t)calibration[Z] - 128) / 256) + 1) * 30;
-
-    ak8963SensorWrite(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_CNTL1, CNTL1_MODE_POWER_DOWN); // power down after reading.
-
-    // Clear status registers
-    ak8963SensorRead(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_ST1, 1, &status);
-    ak8963SensorRead(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_ST2, 1, &status);
-
-    // Trigger first measurement
-    ak8963SensorWrite(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_CNTL1, CNTL1_MODE_ONCE);
-    return true;
-}
+    static ak8963ReadState_e state = CHECK_STATUS;
 
-static bool ak8963Read(int16_t *magData)
-{
     bool ack = false;
-    uint8_t buf[7];
-
-#if defined(MPU6500_SPI_INSTANCE) || defined(MPU9250_SPI_INSTANCE)
 
     // we currently need a different approach for the MPU9250 connected via SPI.
-    // we cannot use the ak8963SensorRead() method for SPI, it is to slow and blocks for far too long.
-
-    static ak8963ReadState_e state = CHECK_STATUS;
+    // we cannot use the ak8963ReadRegisterBuffer() method for SPI, it is to slow and blocks for far too long.
 
     bool retry = true;
```

Comment:
Is used in WAIT_FOR_STATUS condition (23 lines below).

Reply:
Only way to enter `WAITING_FOR_STATUS` is through function call, with `retry == true`. Restart jump goes to `CHECK_STATUS` and it will always `return`.

BTW: simply returning with `state = CHECK_STATUS` and writing the request on next invocation will simplify code. And check failure should be rare, so wasted cycle is not a big problem